File: example/src/main/java/jp/bglb/bonboru/flux/example/middleware/LoadMiddleware.java
package jp.bglb.bonboru.flux.example.middleware;

import jp.bglb.bonboru.flux.action.ActionData;
import jp.bglb.bonboru.flux.example.action.ActionTypes;
import jp.bglb.bonboru.flux.example.dto.MainData;
import jp.bglb.bonboru.flux.example.dto.MainDataBuilder;
import jp.bglb.bonboru.flux.middleware.Middleware;
import jp.bglb.bonboru.flux.store.Store;
import rx.Observable;
import rx.schedulers.Schedulers;

/**
 * Created by Tetsuya Masuda on 2016/07/01.
 */

public class LoadMiddleware extends Middleware<MainData, ActionTypes> {

  @Override public void before(Store<MainData> store) {
    MainData data = store.getData();
    store.setData(new MainDataBuilder().setProgressBarVisibility(true)
        .setError(data.error)
        .setMessage(data.message)
        .setMessages(data.messages)
        .build());
  }

  @Override public Observable<ActionData<MainData, ActionTypes>> intercept(
      Observable<ActionData<MainData, ActionTypes>> observable) {
    return observable.subscribeOn(Schedulers.io());
  }
}

File: kotlinexample/src/main/kotlin/com/example/kotlinexample/middleware/LoadMiddleware.kt
package com.example.kotlinexample.middleware

import com.example.kotlinexample.action.MyActionType
import com.example.kotlinexample.data.MainData
import com.example.kotlinexample.data.MainDataBuilder
import jp.bglb.bonboru.flux.action.ActionData
import jp.bglb.bonboru.flux.kotlin.middleware.Middleware
import jp.bglb.bonboru.flux.store.Store
import rx.Observable
import rx.schedulers.Schedulers

/**
 * Created by tetsuya on 2016/07/01.
 */
class LoadMiddleware() : Middleware<MainData, MyActionType>() {

  override fun before(store: Store<MainData>) {
    store.data = MainDataBuilder().setProgressVisibility(true).setMesage("waiting...").build()
  }

  override fun intercept(
      observable: Observable<ActionData<MainData, MyActionType>>): Observable<ActionData<MainData, MyActionType>> {
    return observable.subscribeOn(Schedulers.io())
  }
}
